src/UserProvider.php

```diff
@@ -35,14 +43,23 @@ public function loadUserByUsername($username)
                 $event = new UserLoadedEvent($user);
                 $this->dispatcher->dispatch('userbase.user_loaded', $event);
             }
-            return $user;
         } catch (RuntimeException $e) {
             throw new UsernameNotFoundException(
                 "A User named \"{$username}\" cannot be found in Userbase.",
                 null,
                 $e
             );
         }
+
+        if (!$this->roleProvider) {
+            return $user;
+        }
+
+        foreach ($this->roleProvider->getRoles($user) as $roleName) {
+            $user->addRole($roleName);
+        }
+
+        return $user;
```

Comment:
*If* a role provider has been injected, pass the loaded User to its getRoles method and add the roles to the User.